quinn-proto/src/connection.rs

```diff
@@ -1258,71 +1205,48 @@ impl Connection {
-                                    reason: TransportError::PROTOCOL_VIOLATION.into(),
-                                },
-                            ));
-                            return State::handshake_failed(
-                                TransportError::PROTOCOL_VIOLATION,
-                                None,
-                            );
+                            return Err(TransportError::PROTOCOL_VIOLATION.into());
                         }
                         while payload.has_remaining() {
                             let version = payload.get::<u32>().unwrap();
                             if version == VERSION {
                                 // Our version is supported, so this packet is spurious
-                                return State::Handshake(state);
+                                return Ok(State::Handshake(state));
                             }
                         }
                         debug!(ctx.log, "remote doesn't support our version");
-                        ctx.events.push_back((
-                            self.handle,
-                            Event::ConnectionLost {
-                                reason: ConnectionError::VersionMismatch,
-                            },
-                        ));
-                        State::Draining(state.into())
+                        Err(ConnectionError::VersionMismatch)
                     }
                     // TODO: SHOULD buffer these to improve reordering tolerance.
                     Header::Short { .. } => {
                         trace!(ctx.log, "dropping short packet during handshake");
-                        State::Handshake(state)
+                        Ok(State::Handshake(state))
                     }
                 }
             }
             State::Established(mut state) => {
                 let id = self.local_id.clone();
                 if let Header::Long { .. } = packet.header {
                     trace!(ctx.log, "discarding unprotected packet"; "connection" => %id);
-                    return State::Established(state);
+                    return Ok(State::Established(state));
                 }
                 let (payload, number) = match self.decrypt_packet(false, packet) {
                     Ok(x) => x,
                     Err(None) => {
                         trace!(ctx.log, "failed to authenticate packet"; "connection" => %id);
-                        return State::Established(state);
+                        return Ok(State::Established(state));
                     }
                     Err(Some(e)) => {
                         warn!(ctx.log, "got illegal packet"; "connection" => %id);
-                        ctx.events
-                            .push_back((self.handle, Event::ConnectionLost { reason: e.into() }));
-                        return State::closed(e);
+                        return Ok(State::closed(e));
```

Comment:
Why aren't we returning an error here? If we did, we could refactor `decrypt_packet` to return `Result<Option<(Vec<u8>, bool)>` and use `?` here for improved concision.